### PR TITLE
Update to MP Metrics 2.0 and a compatible version of SmallRye Fault Tolerance

### LINKED ISSUE #3195: Update SmallRye Metrics to 2.0
**Description**
Update to latest SmallRye Metrics implementation

**Implementation ideas**
Update the version

### PR DESCRIPTION
**Pull request by machi1990**
Fixes #3195. 

Supersides #2695

@jmartisk @kenfinnigan @gsmet 

Thanks to @jmartisk for the initial work.

### CODE REVIEW

**Comment by gsmet** [comment]
So I know I usually ask for squashing but, in this case, we should really have 2 commits: the one from Jan and yours.

Considering each of you contributed significantly to this PR, it's probably better. And this way, I will be able to see your changes.

If it's too much work, let's leave it at that but if not, it would be better.

File: extensions/smallrye-metrics/deployment/src/main/java/io/quarkus/smallrye/metrics/deployment/SmallRyeMetricsProcessor.java
Comment (kenfinnigan) [inline extensions/smallrye-metrics/deployment/src/main/java/io/quarkus/smallrye/metrics/deployment/SmallRyeMetricsProcessor.java]: Does this result in redefining metrics if there multiple different Metric annotations on the same method or class?

Comment (machi1990) [reply extensions/smallrye-metrics/deployment/src/main/java/io/quarkus/smallrye/metrics/deployment/SmallRyeMetricsProcessor.java]: I am not sure I follow.

By `..result in redefining..` you mean `..result in the metric being registered more than once`?

File: extensions/smallrye-metrics/deployment/src/main/java/io/quarkus/smallrye/metrics/deployment/SmallRyeMetricsProcessor.java
Comment (gsmet) [inline extensions/smallrye-metrics/deployment/src/main/java/io/quarkus/smallrye/metrics/deployment/SmallRyeMetricsProcessor.java]: There's something weird here: if the type is not registered but only some methods are registered, then we should not register metrics for all the methods of the classes.

Or did I miss something?

**Review by gsmet — CHANGES_REQUESTED** [review]
Added a couple of comments inline.

File: extensions/smallrye-metrics/deployment/src/main/java/io/quarkus/smallrye/metrics/deployment/SmallRyeMetricsAnnotationsUtils.java
Comment (machi1990) [reply extensions/smallrye-metrics/deployment/src/main/java/io/quarkus/smallrye/metrics/deployment/SmallRyeMetricsAnnotationsUtils.java]: Done

File: extensions/smallrye-metrics/deployment/src/main/java/io/quarkus/smallrye/metrics/deployment/SmallRyeMetricsProcessor.java
Comment (machi1990) [reply extensions/smallrye-metrics/deployment/src/main/java/io/quarkus/smallrye/metrics/deployment/SmallRyeMetricsProcessor.java]: > ..if the type is not registered but only some methods are registered, then we should not register metrics for all the methods of the classes.

I think this is done in `MetricResolver.class` 
```
 private <T extends Annotation> MetricResolver.Of<T> beanResolverOf(MemberInfo element, Class<T> metric, BeanInfo bean) {
        if (bean.isAnnotationPresent(metric)) {
            AnnotationInfo annotation = bean.getAnnotation(metric);
            String name = this.metricName(bean, element, metric, annotation.name(), annotation.absolute());
            Tag[] tags = this.metricTags(annotation);
            return new MetricResolver.DoesHaveMetric(annotation, name, tags);
        } else {
            return (MetricResolver.Of)(bean.getSuperclass() != null ? this.beanResolverOf(element, metric, bean.getSuperclass()) : new MetricResolver.DoesNotHaveMetric());
        }
    }
```

Comment (gsmet) [reply extensions/smallrye-metrics/deployment/src/main/java/io/quarkus/smallrye/metrics/deployment/SmallRyeMetricsProcessor.java]: I'm not sure I have been clear here.

Right now, we do:
- if method annotated, add the class to a set (or whatever structure)
- if class annotated, add the class to a set
- for each class, register metrics for all the methods

Wouldn't it be better to do:
- if method annotated, add the method to a set
- if class annotated, add the class to a set
- for each method which is not part of a registered class, register metrics for the method
- for each class, register metrics for all the methods of the class

The code you point out makes me think we should also consider all the subclasses of the annotated classes (only when the annotation is defined on the class AFAICS).

Comment (machi1990) [reply extensions/smallrye-metrics/deployment/src/main/java/io/quarkus/smallrye/metrics/deployment/SmallRyeMetricsProcessor.java]: Thanks for the clarification. 

My understanding was that the above logic was handled in the resolver code I pointed above and in `Metadata.registerMetrics()` which checks if any metric is present before registering a particular metric. 

But maybe it is better to explicit this logic in the processor as you suggested. 
 
> The code you point out makes me think we should also consider all the subclasses of the annotated classes (only when the annotation is defined on the class AFAICS).

+1 good catch.

Comment (machi1990) [reply extensions/smallrye-metrics/deployment/src/main/java/io/quarkus/smallrye/metrics/deployment/SmallRyeMetricsProcessor.java]: Suggestion taken into account. WDYT @gsmet ?

File: extensions/smallrye-metrics/deployment/src/main/java/io/quarkus/smallrye/metrics/deployment/SmallRyeMetricsProcessor.java
Comment (gsmet) [inline extensions/smallrye-metrics/deployment/src/main/java/io/quarkus/smallrye/metrics/deployment/SmallRyeMetricsProcessor.java]: Maybe a `HashSet` would be better than a list?

File: extensions/smallrye-metrics/deployment/src/main/java/io/quarkus/smallrye/metrics/deployment/SmallRyeMetricsProcessor.java
Comment (gsmet) [inline extensions/smallrye-metrics/deployment/src/main/java/io/quarkus/smallrye/metrics/deployment/SmallRyeMetricsProcessor.java]: I think you should just do:
```
collectedMetricsClasses.put(subClass.name(), subClass);
```
You don't need to go recursive as `index.getAllKnownSubclasses(clazz.name())` will return all the subclasses, not just the direct ones.

**Review by gsmet — CHANGES_REQUESTED** [review]
@machi1990 I think we're nearly there. I added two comments, one minor, the other one not so much.

File: extensions/smallrye-metrics/deployment/src/main/java/io/quarkus/smallrye/metrics/deployment/SmallRyeMetricsProcessor.java
Comment (machi1990) [reply extensions/smallrye-metrics/deployment/src/main/java/io/quarkus/smallrye/metrics/deployment/SmallRyeMetricsProcessor.java]: MethodInfo has no `hashCode` method

File: extensions/smallrye-metrics/deployment/src/main/java/io/quarkus/smallrye/metrics/deployment/SmallRyeMetricsProcessor.java
Comment (machi1990) [reply extensions/smallrye-metrics/deployment/src/main/java/io/quarkus/smallrye/metrics/deployment/SmallRyeMetricsProcessor.java]: Good to know. Thanks.

File: extensions/smallrye-metrics/deployment/src/main/java/io/quarkus/smallrye/metrics/deployment/SmallRyeMetricsProcessor.java
Comment (machi1990) [reply extensions/smallrye-metrics/deployment/src/main/java/io/quarkus/smallrye/metrics/deployment/SmallRyeMetricsProcessor.java]: Sure, thanks. I thought `MethodInfo` has no `hashCode` method just like `ClassInfo`. Will change.

**Review by gsmet — APPROVED** [review]
Looks good to me now.

I will let @jmartisk do a review when he's back but we can already commit that.

Thanks @machi1990  and @jmartisk .

Let's wait for CI before merging.

**Comment by jmartisk** [comment]
Just got back from PTO. Looks nice! Thanks @machi1990 for finishing this.

**Pull request merged**
